crud/schema.go

```diff
@@ -0,0 +1,214 @@
+type Schema map[string]SpaceSchema
+
+// DecodeMsgpack provides custom msgpack decoder.
+func (schema *Schema) DecodeMsgpack(d *msgpack.Decoder) error {
+	var l int
+
+	code, err := d.PeekCode()
+	if err != nil {
+		return err
+	}
+
+	if msgpackIsArray(code) {
+		// Process empty schema case.
+		l, err = d.DecodeArrayLen()
+		if err != nil {
+			return err
+		}
+		if l != 0 {
+			return fmt.Errorf("expected map or empty array, got non-empty array")
+		}
+		*schema = make(map[string]SpaceSchema, l)
+	} else if msgpackIsMap(code) {
+		l, err := d.DecodeMapLen()
+		if err != nil {
+			return err
+		}
+		*schema = make(map[string]SpaceSchema, l)
+
+		for i := 0; i < l; i++ {
+			key, err := d.DecodeString()
+			if err != nil {
+				return err
+			}
+
+			var spaceSchema SpaceSchema
+			if err := d.Decode(&spaceSchema); err != nil {
+				return err
+			}
+
+			(*schema)[key] = spaceSchema
+		}
+	} else {
+		return fmt.Errorf("unexpected code=%d decoding map or empty array", code)
+	}
+
+	return nil
+}
+
+// SpaceSchema contains a single CRUD space schema definition.
+type SpaceSchema struct {
+	Format  []FieldFormat    `msgpack:"format"`
+	Indexes map[uint32]Index `msgpack:"indexes"`
+}
+
+// Index contains a CRUD space index definition.
+type Index struct {
+	Id     uint32      `msgpack:"id"`
+	Name   string      `msgpack:"name"`
+	Type   string      `msgpack:"type"`
+	Unique bool        `msgpack:"unique"`
+	Parts  []IndexPart `msgpack:"parts"`
+}
+
+// IndexField contains a CRUD space index part definition.
+type IndexPart struct {
+	Fieldno     uint32 `msgpack:"fieldno"`
+	Type        string `msgpack:"type"`
+	ExcludeNull bool   `msgpack:"exclude_null"`
+	IsNullable  bool   `msgpack:"is_nullable"`
+}
+
+// SchemaResult contains a schema request result for all spaces.
+type SchemaResult struct {
+	Value Schema
+}
+
+// DecodeMsgpack provides custom msgpack decoder.
+func (schemaResult *SchemaResult) DecodeMsgpack(d *msgpack.Decoder) error {
+	arrLen, err := d.DecodeArrayLen()
+	if err != nil {
+		return err
+	}
+
+	if arrLen == 0 {
+		return fmt.Errorf("unexpected empty response array")
+	}
+
+	// DecodeMapLen inside Schema decode processes `nil` as zero length map,
+	// so in `return nil, err` case we don't miss error info.
+	// https://github.com/vmihailenco/msgpack/blob/3f7bd806fea698e7a9fe80979aa3512dea0a7368/decode_map.go#L79-L81
+	if err = d.Decode(&schemaResult.Value); err != nil {
+		return err
+	}
+
+	if arrLen > 1 {
+		var crudErr *Error = nil
+
+		if err := d.Decode(&crudErr); err != nil {
+			return err
+		}
+
+		if crudErr != nil {
+			return crudErr
+		}
+	}
+
+	for i := 2; i < arrLen; i++ {
+		if err := d.Skip(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// SchemaResult contains a schema request result for a single space.
+type SpaceSchemaResult struct {
+	Value SpaceSchema
+}
+
+// DecodeMsgpack provides custom msgpack decoder.
+func (spaceSchemaResult *SpaceSchemaResult) DecodeMsgpack(d *msgpack.Decoder) error {
```

Comment:
It's better to use a small receiver name in Go (even 1 letter, just `r` is fine).
```suggestion
// DecodeMsgpack provides custom msgpack decoder.
func (result *SpaceSchemaResult) DecodeMsgpack(d *msgpack.Decoder) error {
```

Reply:
Updated